table.go

```diff
@@ -57,6 +60,48 @@ func NewRoutingTable(bucketsize int, localID ID, latency time.Duration, m peerst
 	return rt
 }
 
+func (rt *RoutingTable) GetAllBuckets() []*Bucket {
+	rt.tabLock.RLock()
+	defer rt.tabLock.RUnlock()
+	return rt.Buckets
+}
+
+func (rt *RoutingTable) GenRandPeerID(bucketID int) (peer.ID, error) {
+	rt.tabLock.RLock()
+	bucketLen := len(rt.Buckets)
+	rt.tabLock.RUnlock()
+
+	var targetCpl int
+	if bucketID >= bucketLen-1 {
+		targetCpl = bucketLen
+	} else {
+		targetCpl = bucketID
+	}
+
+	// should give up after a fixed number of attempts so we don't take up too much time/cpu
+	for i := 0; i < 1000; i++ {
+		peerID, err := randPeerID()
+		if err != nil {
+			log.Debugf("failed to generate random peerID in bucket %d, error is %+v", bucketID, err)
+			continue
+		}
+		if CommonPrefixLen(ConvertPeerID(peerID), rt.local) == targetCpl {
+			return peerID, err
+		}
+	}
```

Comment:
Hm. This is very unfortunate. Given our network size, we should have _at least_ 16 buckets). That's 2**16 hashes to find a peer in that last bucket.

I don't know _how_ to fix this on our current DHT. Given how we've defined the protocol, we _have_ to know the un-hashed key.

Honestly, I think the best solution is to pre-compute a table with a few known-good bootstrap keys in each of the first 20 buckets. If you can write a program to do this (ideally highly parallel), I can run it on a really beefy machine for a few days.

Reply:
I'm probably missing context here, but a couple comments:
1) Why do we need a public non-test function for computing a random peer ID?
2) Even if we needed to generate random peer IDs we don't need to hash anything since the `randPeerID` function does not return the material that gets hashed into the PeerID. As a result, if I have `n` buckets and want bucket `i` that has 20 matching prefix bits with me all I should need to do is randomly generate a single number from 0 to 2^(n-20) and then just prepend the matching 20 bits. So very little computation (one random number generation) should be required.

Reply:
@aschmahmann this is for https://github.com/libp2p/go-libp2p-kad-dht/issues/375 (I've updated the PR description).

> 2

Due to how _our_ DHT works, the we need to send the _unhashed_ key with the request. That means we need to know the unhashed key. Kademlia as described by the paper assumes that all keys will be pre-hashed. If we wanted to do that, we'd have to pre-hash the keys when we store records in the datastore (e.g., peer routing information).

We _could_ modify the DHT to send _both_ the unhashed and hash keys but that seems like a lot of extra work just to support bootstrapping.

Reply:
@Stebalien thx for the PR pointer. Fair enough about not wanting to do much changing to the dht message types if we don't have to. However, I'm not sure the solution of pre-generating keys helps. We can pre-generate keys for a single peer, but generating them for all peers is going to need require packaging a ton of keys (because the bucket distance is computed based on your local ID).

In any event it still doesn't seem like we need the hash function inside of `randPeerID`, just the one in `ConvertPeerID`.

Reply:
Ah, yeah, now I feel like an idiot. This isn't going to work (we'd need to ship ~1m peer IDs).

> In any event it still doesn't seem like we need the hash function inside of randPeerID, just the one in ConvertPeerID.

The issue is that we need to make a DHT request with this key. To make the DHT request, we need to know the non-hashed key.

Reply:
Turns out this didn't need a beefy machine and we really don't need 1m peer IDs. I've created a branch with a table mapping 20 bit hash prefixes to numbers that can be used to generate peer IDs that hash to these 20 bit hash prefixes: https://github.com/libp2p/go-libp2p-kbucket/tree/feat/bucket-prefix-gen

More concretely, to find a peer ID in a bucket N, we'll have to:

1. Generate a random 20 bit number.
2. Hash our peer ID (sha256)
3. Replace the first N bits with the first N bits from our hashed peer ID.
4. Lookup this random number in the table in the table to get SEED.
5. Encode SEED into a 32 byte slice using BigEndian ordering.
6. Use this slice as the hash digest in a sha256 multihash.
7. This multihash is the peer ID.

When this peer ID is hashed, the first N bits should match the result of step (2).

Reply:
The downside is that this increases our binary size by 4MiB. But I'm not _that_ concerned.

Reply:
I've reduced it to 16 bits (16 buckets) to reduce the size to 256KiB.

Reply:
@Stebalien  This was a really cool idea ! Have rebased on top of your branch & made all the changes. Let me know what you think.

PS: Made some minor changes to generate/main.go(line 31 & 58) & generated a new prefix-map to make it easier to use. 

Basically, instead of constructing the multi-hash like this :
```
[]byte{mh.Sha256,31, 24 zero byte slice ,8 bytes of Uint64}
```

We now construct it as
```
[]byte{mh.Sha256,31, 24 zero byte slice ,8 bytes of Uint64}
```

Reply:
Also, apologies for the delay in getting this up. Things were a bit hectic at work.